src/rust/mqtt-publisher/src/main.rs

```diff
@@ -0,0 +1,230 @@
+    order_id: BigDecimal,
+    direction: bool,
+    integrator: String,
+    self_match_behavior: i16,
+    size: BigDecimal,
+}
+#[derive(Serialize, Deserialize)]
+struct PlaceSwapOrderNotif {
+    txn_version: BigDecimal,
+    event_idx: BigDecimal,
+    time: DateTime<Utc>,
+    market_id: BigDecimal,
+    order_id: BigDecimal,
+    direction: bool,
+    signing_account: String,
+    integrator: String,
+    min_base: BigDecimal,
+    max_base: BigDecimal,
+    min_quote: BigDecimal,
+    max_quote: BigDecimal,
+    limit_price: BigDecimal,
+}
+#[derive(Serialize, Deserialize)]
+struct ChangeOrderSizeNotif {
+    txn_version: BigDecimal,
+    event_idx: BigDecimal,
+    time: DateTime<Utc>,
+    market_id: BigDecimal,
+    user: String,
+    custodian_id: BigDecimal,
+    order_id: BigDecimal,
+    side: bool,
+    new_size: BigDecimal,
+}
+#[derive(Serialize, Deserialize)]
+struct CancelOrderNotif {
+    txn_version: BigDecimal,
+    event_idx: BigDecimal,
+    time: DateTime<Utc>,
+    market_id: BigDecimal,
+    user: String,
+    custodian_id: BigDecimal,
+    order_id: BigDecimal,
+    reason: i16,
+}
+#[derive(Serialize, Deserialize)]
+struct FillNotif {
+    txn_version: BigDecimal,
+    event_idx: BigDecimal,
+    emit_address: String,
+    time: DateTime<Utc>,
+    maker_address: String,
+    maker_custodian_id: BigDecimal,
+    maker_order_id: BigDecimal,
+    maker_side: bool,
+    market_id: BigDecimal,
+    price: BigDecimal,
+    sequence_number_for_trade: BigDecimal,
+    size: BigDecimal,
+    taker_address: String,
+    taker_custodian_id: BigDecimal,
+    taker_order_id: BigDecimal,
+    taker_quote_fees_paid: BigDecimal,
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let mqtt_url = std::env::var("MQTT_URL")?;
+    let db_url = std::env::var("DATABASE_URL")?;
+
+    let mut mqttoptions = MqttOptions::parse_url(mqtt_url).unwrap();
+    mqttoptions.set_transport(Transport::Tcp);
+    mqttoptions.set_keep_alive(Duration::from_secs(5));
+    let (client, mut eventloop) = AsyncClient::new(mqttoptions, 10);
+
+    let mut listener = PgListener::connect(&db_url).await?;
+
+    let channels = vec![
+        "place_limit_order",
+        "place_market_order",
+        "place_swap_order",
+        "change_order_size",
+        "cancel_order",
+        "fill",
+    ];
+
+    listener.listen_all(channels).await?;
+
+    tokio::task::spawn(async move {
+        loop {
+            eventloop.poll().await.unwrap();
+        }
+    });
+
+    loop {
+        let notification = listener.recv().await?;
+        match notification.channel() {
+            "place_limit_order" => {
+                let data: PlaceLimitOrderNotif = serde_json::from_str(notification.payload())?;
+                client
+                    .publish(
+                        format!(
+                            "place_limit_order/{}/{}/{}/{}",
+                            data.market_id, data.user, data.custodian_id, data.integrator
```

Comment:
I'm assuming that these fields are just optionally filterable, and in any combination, right?

For example I can filter on everything except `integrator` to get all my place limit order events, regardless of the integrator

Also, what if I want all my limit orders (unique user/integrator) for all markets (no market specified)?

Reply:
Yes.